### PR TITLE
Add templates for all common checks in golang repos

### PR DESCRIPTION
**Pull request by lluiscampos**
This commit adds the following templates:
* .gitlab-ci-check-golang-static.yml: Static checks like format, gocyclo
* .gitlab-ci-check-golang-unittests.yml: Run unit tests and publish
results into codecov
* .gitlab-ci-check-golang-dockerimage.yml: Build Docker image and
publish it into Docker Hub
* .gitlab-ci-check-apidocs.yml: Check Swagger formatted docs

They cover all common jobs for our Go repos.

### CODE REVIEW

**Comment by lluiscampos** [comment]
See demo usage in https://github.com/mendersoftware/workflows/pull/9

File: .gitlab-ci-check-apidocs.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# .gitlab-ci-check-apidocs.yml
+#
+# This gitlab-ci template validates Swawer API specification assuming they
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply .gitlab-ci-check-apidocs.yml]: :+1:

File: .gitlab-ci-check-golang-dockerimage.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
+# .gitlab-ci-check-golang-dockerimage.yml
+#
+# This gitlab-ci template builds a Docker image assuming a single Dockerfile
+# and publish is into Docker Hub
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply .gitlab-ci-check-golang-dockerimage.yml]: :ok_hand:

File: .gitlab-ci-check-golang-static.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    - go get -u github.com/fzipp/gocyclo
+    # Prepare GOPATH for the build
+    - mkdir -p /go/src/github.com/mendersoftware
+    - cp -r ${CI_PROJECT_DIR} /go/src/github.com/mendersoftware/${CI_PROJECT_NAME}
+    - cd /go/src/github.com/mendersoftware/${CI_PROJECT_NAME}
+  script:
+    # Test if code was formatted with 'go fmt'
+    # Command will format code and return modified files
+    # fail if any have been modified.
+    - if [ -n "$(go fmt)" ]; then echo 'Code is not formatted with "go fmt"'; false; fi
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply .gitlab-ci-check-golang-static.yml]: I haven't verified locally but I assume not: this is the exact command line call used currently in all our repos.

**Comment by lluiscampos** [comment]
> I like this a lot! Only nitpicks to go.
I fixed the nitpicks, thanks!

What about the names? I was expecting some comments from you:
- `golang-static`: I decided to create a single one rather than one for format and other for gocyclo/vet. This breaks your original idea with python3-format... But I think it makes sense
- `dockerimage` and `unittests` or something like `docker-image-build-and-publish` and `unit-tests-run-and-coverage-report`?

> I think you might be in danger of making yourself superfluous soon :p

Thank you? ;)

File: .gitlab-ci-check-golang-dockerimage.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+build:
+  stage: build
+  tags:
+    - docker
+  image: docker
+  services:
+    - docker:dind
+  before_script:
+    - export DOCKER_REPOSITORY="mendersoftware/${CI_PROJECT_NAME}"
```
Comment (mzedel) [inline .gitlab-ci-check-golang-dockerimage.yml]: very nice! didn't think of it at the time and assumed the project names would diverge

**Comment by oleorhagen** [comment]
> What about the names? I was expecting some comments from you:
> 
>     * `golang-static`: I decided to create a single one rather than one for format and other for gocyclo/vet. This breaks your original idea with python3-format... But I think it makes sense
> 
>     * `dockerimage` and `unittests` or something like `docker-image-build-and-publish` and `unit-tests-run-and-coverage-report`?
> 

Hmm, yeh. I noticed you collected them, but I don't think this is an issue really. To be honest, I think the static keyword is better, as it collects...

Hmm, on the other hand. As I'm typing this I kinda wanna object. It is nice to know, from simply looking at the pipeline, what was actually wrong without having to checkout the runner's output. Then on the other hand, this will create a lot of runners.

I guess I'm kinda in favour of the `language-task` here, without collecting them. But not so strongly I will vote this down. I guess that means, it's up to you here.

**Comment by lluiscampos** [comment]
> > What about the names? I was expecting some comments from you:
> > ```
> > * `golang-static`: I decided to create a single one rather than one for format and other for gocyclo/vet. This breaks your original idea with python3-format... But I think it makes sense
> > ```
> 
> Hmm, yeh. I noticed you collected them, but I don't think this is an issue really. To be honest, I think the static keyword is better, as it collects...
> 
> Hmm, on the other hand. As I'm typing this I kinda wanna object. It is nice to know, from simply looking at the pipeline, what was actually wrong without having to checkout the runner's output. Then on the other hand, this will create a lot of runners.
> 
> I guess I'm kinda in favour of the `language-task` here, without collecting them. But not so strongly I will vote this down. I guess that means, it's up to you here.

I've been thinking back and forth about this one. I ultimately decided on keeping them collected on "static". I think that splitting them will risk some repos excluding some of the checks by mistake (and also it will consume more CI minutes).

> > ```
> > * `dockerimage` and `unittests` or something like `docker-image-build-and-publish` and `unit-tests-run-and-coverage-report`?
> > ```
> 
> I would be in favour of the last one here, even though the names are becoming excruciatingly long, they are very descriptive, and we're not in 70's unix land. But that's just my opinion here.

And here I also concluded to leave it as is. When I look at the full list of files, I feel like they will differ too much from other checks (plus getting filenames extremely long).

**Comment by lluiscampos** [comment]
@oleorhagen Regardless of the conclusions, thank you very much for your input!

**Pull request merged**